docs/packaging-your-sql-project/intro-to-python-scripting-and-pipelines.md

```diff
@@ -11,4 +11,143 @@ kernelspec:
+  Keyword arguments:
+  real -- the real part (default 0.0)
+  imag -- the imaginary part (default 0.0)
+  """
+  if imag == 0.0 and real == 0.0:
+      return complex_zero
+  ...
+```
+
+#### Modularization
+
+- Modularization is the technique of breaking down your script into smaller, self-contained modules or functions. Each module or function performs a specific task, making the code more readable, maintainable, and reusable.
+- Moreover, modularization enhances collaboration as multiple developers can work on different modules simultaneously, promoting collaboration and parallel development. This speeds up the overall development process and allows teams to work on specific parts of the project independently.
+- Modularization is often used in conjunction with the main program, which calls the functions defined in the modules to execute the main logic of the script. This is discussed in more detail below.
+
+### The Main Program
+
+#### What does `if __name__ == "__main__:"` mean?
+
+Typically, a script will have a `main` function that calls other functions to execute the main logic of the script. The `main` function is called from the `if __name__=="__main__":` block, which is executed when the script is run directly (not imported as a module). This allows you to have reusable code in your script that can be imported into other scripts without unintended side effects.
+
+Inside the `if __name__=="__main__":` block, you can call the functions you defined earlier to execute the main logic of your script:
+
+```python
+if __name__=="__main__":
+    # call functions here
+    pass
+```
+
+#### Calling Functions from the Main Program
+
+Modularized scripts can be called in the main program as follows:
+
+```{code-cell} ipython3
+def add(int1, int2):
+    """This function adds two integers"""
+    return int1 + int2
+
+
+def subtract(int1, int2):
+    """This function subtracts two integers"""
+    return int1 - int2
+
+
+if __name__ == "__main__":
+    num1 = 10
+    num2 = 5
+
+    result_add = add(num1, num2)
+    result_subtract = subtract(num1, num2)
+
+    print(f"Addition: {result_add}")
+    print(f"Subtraction: {result_subtract}")
+```
+
+```{important}
+When you call if `__name__ == "__main__":` in different scripts, it allows you to use the scripts with your functions as reusable modules that can be imported into other scripts. For instance, the above example could be modularized with the functions placed in a separate script called `math_operations.py` and the main logic in an another script called `main.py`. `main.py` would `import math_operations` as a module and call the functions like so: `math_operations.add(num1, num2)`.
+```
+
+## Python Scripting for Data Pipelines
+
+Python scripting is a powerful way to automate data tasks, such as downloading data from the web, cleaning and transforming data, and saving it to a database. The combination of `pandas` and `SQLAlchemy` provides a robust toolkit for working with data in Python.
```

Comment:
Add links to the doc pages of `pandas` and `SQLAlchemy`

Reply:
fixed in latest commit